lib/Cleantalk/Antispam/IntegrationsByClass/WPForms.php

```diff
@@ -0,0 +1,227 @@
+     * @psalm-suppress PossiblyUnusedMethod
+     */
+    public function doPublicWork()
+    {
+        add_action('wpforms_frontend_output', [$this, 'addField'], 1000, 5);
+    }
+
+    public function doAjaxWork()
+    {
+        if (!$this->isSkipIntegration()) {
+            add_filter('wpforms_process_initial_errors', [$this, 'showResponse'], 100, 2);
+        }
+    }
+
+    /**
+     * Inserts anti-spam hidden to WPForms
+     *
+     * @return void
+     * @global State $apbct
+     */
+    public function addField($_form_data, $_some, $_title, $_description, $_errors)
+    {
+        global $apbct;
+
+        ct_add_hidden_fields('ct_checkjs_wpforms');
+        echo Honeypot::generateHoneypotField('wp_wpforms');
+        if ( $apbct->settings['trusted_and_affiliate__under_forms'] === '1' ) {
+            echo Escape::escKsesPreset(
+                apbct_generate_trusted_text_html('label_left'),
+                'apbct_public__trusted_text'
+            );
+        }
+    }
+
+    /**
+     * Adding error to form entry if message is spam
+     * Call spam test from here
+     *
+     * @param array $errors
+     * @param array $form_data
+     *
+     * @return array
+     */
+    public function showResponse($errors, $form_data)
+    {
+        if (!empty($errors)) {
+            return $errors;
+        }
+
+        $spam_comment = $this->testSpam();
+
+        error_log('showResponse');
+        error_log(print_r($spam_comment, true));
+
+        if (!$spam_comment ) {
+            return $errors;
+        }
+
+        $field_id = 0;
+        if ( $form_data && ! empty($form_data['fields']) && is_array($form_data['fields']) ) {
+            foreach ( $form_data['fields'] as $key => $field ) {
+                if ( array_search('email', $field) === 'type' ) {
+                    $field_id = $key;
+                    break;
+                }
+            }
+        }
+
+        $field_id = ! $field_id && $form_data && ! empty($form_data['fields']) && is_array($form_data['fields'])
+            ? key($form_data['fields'])
+            : $field_id;
+
+        if ( isset($form_data['id']) ) {
+            $errors[$form_data['id']][$field_id] = $spam_comment;
+        }
+
+        return $errors;
+    }
+
+    /**
+     * Test WPForms message for spam
+     * Doesn't hooked anywhere.
+     * Called directly from apbct_form__WPForms__showResponse()
+     *
+     * @return string|void
+     * @global State $apbct
+     */
+    public function testSpam()
+    {
+        global $apbct;
+
+        $checkjs = apbct_js_test(Sanitize::cleanTextField(Post::get('ct_checkjs_wpforms')));
+        $input_array = apply_filters('apbct__filter_post', $_POST);
+        $params = ct_gfa_dto($input_array)->getArray();
+
+        if (isset($input_array['wpforms']['fields']) && is_array($input_array['wpforms']['fields']) &&
+            (isset($params['nickname']) && $params['nickname'] === '')
+        ) {
+            $params['nickname'] = '';
```

Comment:
I see that this code chunk was migrated from get fields any, but the new logic is not the same. Let's talk about this on daily scrum.

Reply:
Thanks, fixed https://github.com/CleanTalk/wordpress-antispam/pull/607/commits/cc725b3565ede9ab58f738accc64db2118832629